src/relations/db.py

```diff
@@ -0,0 +1,258 @@
+            )
+            self.charm.unit.status = BlockedStatus("extensions requested through relation")
+            return
+
+        # Sometimes a relation changed event is triggered,
+        # and it doesn't have a database name in it.
+        database = event.relation.data[event.app].get("database")
+        if not database:
+            logger.warning("No database name provided")
+            event.defer()
+            return
+
+        try:
+            # Creates the user and the database for this specific relation if it was not already
+            # created in a previous relation changed event.
+            user = f"relation_id_{event.relation.id}"
+            password = unit_relation_databag.get("password", new_password())
+            self.charm.postgresql.create_user(user, password, self.admin)
+            self.charm.postgresql.create_database(database, user)
+
+            # Build the primary's connection string.
+            primary = str(
+                ConnectionString(
+                    host=self.charm.primary_endpoint,
+                    dbname=database,
+                    port=5432,
+                    user=user,
+                    password=password,
+                    fallback_application_name=event.app.name,
+                )
+            )
+
+            # Build the standbys' connection string.
+            standbys = str(
+                ConnectionString(
+                    host=self.charm.replicas_endpoint,
+                    dbname=database,
+                    port=5432,
+                    user=user,
+                    password=password,
+                    fallback_application_name=event.app.name,
+                )
+            )
+
+            # Set the data in both application and unit data bag.
+            # It 's needed to run this logic on every relation changed event
+            # setting the data again in the databag, otherwise the application charm that
+            # is connecting to this database will receive a "database gone" event from the
+            # old PostgreSQL library (ops-lib-pgsql) and the connection between the
+            # application and this charm will not work.
+            for databag in [application_relation_databag, unit_relation_databag]:
+                # This list of subnets is not being filled correctly yet.
+                databag["allowed-subnets"] = self._get_allowed_subnets(event.relation)
+                databag["allowed-units"] = self._get_allowed_units(event.relation)
+                databag["host"] = self.charm.endpoint
+                databag["master"] = primary
+                databag["port"] = "5432"
+                databag["standbys"] = standbys
+                databag["state"] = self._get_state()
+                databag["version"] = self.charm.postgresql.get_postgresql_version()
+                databag["user"] = user
+                databag["password"] = password
+                databag["database"] = database
```

Comment:
Shayan mentioned a slightly cleaner way of doing this in the pgbouncer charm:
```python
databag = {
    "allowed-subnets": self._get_allowed_subnets(event.relation),
    "allowed-units": self._get_allowed_units(event.relation),
    # ....
}

databag.update(updates)
```

Reply:
Nice suggestion. I implemented it on https://github.com/canonical/postgresql-k8s-operator/pull/11/commits/33c493834552bc49d9c9d3b06a58474c87e61c7e.